dali/kernels/kernel.h

```diff
@@ -0,0 +1,178 @@
+    const InListGPU<Input2, 4> &in2,
+    const InTensorCPU<float, 3> &aux);
+};
+
+}  // namespace examples
+
+/// @brief A collection of pseudo-methods to operate on Kernel classes/objects
+namespace kernel {
+
+// avoid retyping "Kernel" every second word...
+
+template <typename Kernel>
+using inputs = kernel_inputs<Kernel>;
+
+template <typename Kernel>
+using outputs = kernel_outputs<Kernel>;
+
+template <typename Kernel>
+using args = kernel_args<Kernel>;
+
+using Context = KernelContext;
+using Requirements = KernelRequirements;
+
+/// @brief Gets requirements for given Kernel
+/// @param context            - execution environment (without scratch memory)
+/// @param input              - kernel inputs, convertible to kernel_inputs<Kernel>
+/// @param args               - kernel extra arguments, convertible to kernel_args<Kernel>
+template <typename Kernel>
+Requirements GetRequirements(
+      Context &context,
+      const inputs<Kernel> &input,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  return apply_all(Kernel::GetRequirements, context, input, args);
+}
+
+/// @brief Executes a Kernel on an input set
+/// @param context             - execution environment (with scratch memory)
+/// @param input               - kernel inputs, convertible to kernel_inputs<Kernel>
+/// @param outputs             - kernel outputs, convertible to kernel_outputs<Kernel>
+/// @param args                - kernel extra arguments, convertible to kernel_args<Kernel>
+template <typename Kernel>
+void Run(
+      Context &context,
+      const outputs<Kernel> &output,
+      const inputs<Kernel> &input,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  apply_all(Kernel::Run, context, output, input, args);
+}
+
+/// @brief Default implementation of requirements for multiple input sets
+/// @remarks Reuse scratch, append output shapes
+///
+/// @TODO(michalz) remove references from inputs/args?
+template <typename Kernel>
+Requirements GetRequirements(
+      Context &context,
+      const std::vector<inputs<Kernel>> &input_sets,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  if (input_sets.empty())
+    return {};
+
+  Requirements req = GetRequirements<Kernel>(context, input_sets[0], args);
+  for (size_t i = 1; i < input_sets; i++) {
+    Requirements newReq = GetRequirements<Kernel>(context, input_sets[i], args);
+    req.AddInputSet(newReq, true);
+  }
+  return req;
+}
+
+/// @brief Default implementation of execution for multiple input sets
+///
+/// @TODO(michalz) remove references from inputs/outputs/args?
+template <typename Kernel>
+void Run(Context &context,
+      const std::vector<outputs<Kernel> > &output_sets,
+      const std::vector<inputs<Kernel> > &input_sets,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  assert(output_sets.size() == input_sets.size());
+  for (size_t i = 0; i < input_sets.size(); i++) {
+    Run<Kernel>(context, output_sets, input_sets, args);
```

Comment:
Shouldn't it be something like 
```suggestion
    Run<Kernel>(context, output_sets[i], input_sets[i], args);
```